storage: implement simple version of souce-specific statistics (#16788)
MaterializeInc/materialize: doc/user/content/sql/system-catalog/mz_internal.md

`cpu_nano_cores`   | [`bigint`] | Approximate CPU usage, in billionths of a vCPU core.
`memory_bytes`     | [`bigint`] | Approximate RAM usage, in bytes.

### `mz_storage_source_statistics`

The `mz_storage_source_statistics` table gives the last known storage-specific statistics

> for all processes of all extant storage replicas.

Reviewer: I think we said we didn't want to use the term "replica" for these. For now, it seems we can stick to "storage hosts" for this.

Reviewer: Or just use sth like "extant sources and sinks"

Author: I copied this from `mz_storage_host_metrics` above haha...will change them both

At this time, we do not make any guarantees about the exactness or freshness of these numbers.

Field                 | Type        | Meaning
----------------------|-------------|--------
`id`                 | [`text`]     | The ID of the storage object (source or sink).
`worker_id`          | [`bigint`]   | An identifier of a process within a host.
`snapshot_committed` | [`boolean`]  | Whether or not we have committed the initial snapshot for a source.
`messages_received`  | [`bigint`]   | How many messages we have received from upstream.
`messages_committed` | [`bigint`]   | How many messages we have committed into our storage layer.
`bytes_received`     | [`bigint`]   | How many bytes worth of messages we have received from upstream.

> These values all require some amount of interpretation:

Reviewer: Very thoughtful! \o/

let's talk in person but I think this is close to what we want. Or it!

- `messages_received` does not distinguish between inserts and deletes. Some upstream sources, like postgres,
  produce deletion messages that we count the same as insertions.
- `messages_committed` can be permanently lower or higher than `messages_received`, because:

>   - The storage layer can choose to consolidate messages within 1 second

Reviewer: nit: one second happens to be the current (default) setting but it might change, so maybe just talk about "within a certain time interval"?

Author: will change this!

  - Some envelopes (like `UPSERT`) can produce 2 messages (an insert and a delete) from a single upstream message.
- `bytes_recieved` can represent different things for different sources.
  - For example, it represents the literal size of the key and value from kafka messages, but represents
  an internal representation of rows for postgres sources.

> - The `messages_` and `bytes_received` metrics can be interpreted as rates that can loosely show

Reviewer: typo: `messages_*` ?

`cpu_nano_cores`   | [`bigint`] | Approximate CPU usage, in billionths of a vCPU core.
`memory_bytes`     | [`bigint`] | Approximate RAM usage, in bytes.

### `mz_source_statistics`

The `mz_storage_source_statistics` table contains statistics for each process of each source in the system.

Materialize does not make any guarantees about the exactness or freshness of these numbers. As internal
components of the system are restarted, some of these metrics will be reset to 0.

Field                 | Type         | Meaning
----------------------|--------------|--------

> `id`                  | [`text`]     | The ID of the source.

Reviewer 2: ```suggestion
`id`                  | [`text`]     | The ID of the source. Corresponds to [mz_catalog.mz_sources.id](https://materialize.com/docs/sql/system-catalog/mz_catalog#mz_sources).
```

> `worker_id`           | [`bigint`]   | An identifier of a worker within the source.

Reviewer 2: ```suggestion
`worker_id`           | [`bigint`]   | The ID of the worker thread.
```

`snapshot_committed`  | [`boolean`]  | Whether or not the worker has committed the initial snapshot for a source.

> `messages_received`   | [`bigint`]   | How many raw messages the worker has received from upstream.
> `updates`             | [`bigint`]   | How many updates (inserts + deletes) the worker has written but not yet committed to the storage layer.
> `updates_committed`   | [`bigint`]   | How many updates (inserts + deletes) the worker has committed into our storage layer.
> `bytes_received`      | [`bigint`]   | How many bytes worth of messages the worker has received from upstream. The way the bytes are counted is source-specific.

Reviewer 2: ```suggestion
`messages_received`   | [`bigint`]   | The number of raw messages the worker has received from upstream.
`updates`             | [`bigint`]   | The number of updates (inserts + deletes) the worker has written but not yet committed to the storage layer.
`updates_committed`   | [`bigint`]   | The number of updates (inserts + deletes) the worker has committed into the storage layer.
`bytes_received`      | [`bigint`]   | The number of bytes worth of messages the worker has received from upstream. The way the bytes are counted is source-specific.
```

These values all require some amount of interpretation:

> - `messages_received` does not distinguish between inserts and deletes. Some upstream sources, like postgres,
>   produce deletion messages that are counted the same as insertions.
> - The `updates*` metrics can be permanently lower or higher than `messages_received`, because:
>   - The storage layer can choose to consolidate messages within a certain time interval.
>   - Some envelopes (like `UPSERT`) can produce 2 messages (an insert and a delete) from a single upstream message.
> - The `messages_*` and `bytes_received` metrics can optionally be interpreted as rates that can loosely show
>   the changing speed of a source over time.
> - In general, the values should be summed across `worker_id`'s, and some metrics may only exist for some workers.
>   - Additionally, sources with multiple outputs (like postgres shards) will need `messages_committed`
>   and `snapshot_committed` summed/AND-ed across the top-level source and the connected tables.

Reviewer 2: ```suggestion
- `messages_received` does not distinguish between inserts and deletes. Some upstream sources, like PostgreSQL,
  produce deletion messages that are counted the same as insertions.
- The `updates*` metrics can be permanently lower or higher than `messages_received`, because:
  - The storage layer can choose to consolidate messages within a certain time interval.
  - Some envelopes (like `UPSERT`) can produce 2 messages (an insert and a delete) from a single upstream message.
- The `messages_*` and `bytes_received` metrics can optionally be interpreted as rates that can loosely show
  the changing speed of a source over time.
- In general, the values should be summed across `worker_id`s, and some metrics may only exist for some workers.
  - Additionally, sources with multiple outputs (like PostgreSQL) will need `messages_committed`
  and `snapshot_committed` summed/AND-ed across the top-level source and its subsources.
```

Reviewer 2: It's reasonable to keep these details here, for lack of a better place (at least for the time being).

`cpu_nano_cores`   | [`bigint`] | Approximate CPU usage, in billionths of a vCPU core.
`memory_bytes`     | [`bigint`] | Approximate RAM usage, in bytes.

> ### `mz_storage_source_statistics`

Reviewer 3: "storage source" is not the right name! Either "storage host" or "source". But I think having two separate views—`mz_source_statistics` and `mz_sink_statistics`—is best, since we plan to keep the general structure of these views even after transitioning them to introspection views, and then they'll be consistent with e.g. `mz_{source,sink}_status`.

> The `mz_storage_source_statistics` table gives the last known storage-specific statistics
> for all processes of all extant storage sources.

Reviewer 3: ```suggestion
The `mz_storage_source_statistics` table contains statistics for each process of each source in the system.
```

> At this time, we do not make any guarantees about the exactness or freshness of these numbers.

Reviewer 3: ```suggestion
Materialize does not make any guarantees about the exactness or freshness of these numbers.
```

Reviewer 3: I think this is the place to mention that these statistics periodically reset to 0 as internal components of the system restart.

Field                 | Type        | Meaning
----------------------|-------------|--------

> `id`                 | [`text`]     | The ID of the storage object (source or sink).

Reviewer 3: ```suggestion
`id`                 | [`text`]     | The ID of the source.
```

> `worker_id`          | [`bigint`]   | An identifier of a process within a host.

Reviewer 3: ```suggestion
`worker_id`          | [`bigint`]   | An identifier of a worker within the source.
```

`snapshot_committed` | [`boolean`]  | Whether or not we have committed the initial snapshot for a source.
`messages_received`  | [`bigint`]   | How many messages we have received from upstream.
`messages_committed` | [`bigint`]   | How many messages we have committed into our storage layer.

> `bytes_received`     | [`bigint`]   | How many bytes worth of messages we have received from upstream.

Reviewer 3: It's not clear who "we" is in these descriptions. The worker? The process? The source as a whole? The descriptions of these fields in the design doc used a more precise noun in place of "we" (e.g., "the worker").

These values all require some amount of interpretation:

- `messages_received` does not distinguish between inserts and deletes. Some upstream sources, like postgres,
  produce deletion messages that we count the same as insertions.
- `messages_committed` can be permanently lower or higher than `messages_received`, because:
  - The storage layer can choose to consolidate messages within a certain time interval.

>   - Some envelopes (like `UPSERT`) can produce 2 messages (an insert and a delete) from a single upstream message.

Reviewer 3: I think we need to go back to the drawing board on this. IMO we can't have `messages_received` and `messages_committed` count messages differently, as they both use "messages" in their name. The metric as described is, I think, better named something like `updates_committed`... but I don't think that's necessarily useful to expose? (In the future, I think `updates_committed` is computable with `SELECT count(*) FROM changes(src)`, and in the meantime I think `SELECT count(*) FROM src` is a decent proxy for `updates_committed`.)

Put another way: the value in exposing these two quantities seemed to be that you could watch for skew between `messages_received and `messages_committed` to detect problems in the source ingestion pipeline. If these two metrics are expected to skew, they seem substantially less useful, and at the very least seems like we need to rename them so that they do not seem so comparable.

Reviewer: strawman: we could have three statistics:
 - `messages_received`: the raw messages that we get from the source
 - `updates` (name tbd) number of updates that we have _after_ messages have gone through the enveloping logic
 - `updates_committed`
 
Depending on the envelope, there will be systemic skew between `messages_received` and `updates`, but with a constant factor/bound. On the other hand,  `updates` and `updates_committed` should proceed in lockstep and you will notice that something is afoot when they don't. wdyt?

Reviewer: cc my usual crew for these user-facing things @morsapaes, @sjwiesman, @joacoc, @ggnall  👋

Reviewer: cc my usual crew for these user-facing things @morsapaes, @sjwiesman, @joacoc, @ggnall  👋

Reviewer 4: I think @aljoscha 's take makes sense. Idk what the right names are, I'll need to think about it.

Author: @aljoscha `updates` can be incremented when we write the updates into (non-committed) persist batches, so it will always move in lock-step with `updates_committed`

Sounds like a reasonable plan to me!

Author: Updated the pr to reflect this (with the given placeholder names)

Reviewer 3: I like it! Maybe `updates_staged` instead of `updates`?

Reviewer 2: +1 for `updates_staged`.

- `bytes_recieved` can represent different things for different sources.
  - For example, it represents the literal size of the key and value from kafka messages, but represents

>   an internal representation of rows for postgres sources.

Reviewer 3: This feels like too much detail. I would rather add a sentence like "The way in which bytes are counted is specific to each source implementation", or something like that, to the table row, and leave it at that.

- The `messages_*` and `bytes_received` metrics can be interpreted as rates that can loosely show
  the changing speed of a source over time. These metrics are periodically reset to 0.
- In general, the values should be summed across `worker_id`'s, and some metrics may only exist for some workers.
  - Additionally, sources with multiple outputs (like postgres shards) will need `messages_committed`

>   and `snapshot_committed` summed/AND-ed across the top-level source and the connected tables.

Reviewer 3: Thanks for taking the first cut at this. Looping in @morsapaes to make sure this goes through a docs style pass. The high-level point: this page, more than almost any other page in our docs, is reference documentation. It should deal almost exclusively in hard facts. We don't want to be giving folks advice about how to use the data here—that's better left to guides and tutorials—we should just describe what the data _is_.

Author: I definitely think this interpretation section is important, but if it should go somewhere else I'll move it!
